include/mbgl/style/expression/match.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+
+#include <mbgl/style/expression/expression.hpp>
+#include <mbgl/style/expression/parsing_context.hpp>
+#include <mbgl/style/conversion.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using InputType = variant<int64_t, std::string>;
+using MatchKey = variant<InputType, std::vector<InputType>>;
+
+template <typename T>
```

Comment:
My feeling is no on the latter, because I would think that any realistic use case for that could be handled by simply coercing the input to a string and then matching on string keys. 

If users _really_ want to distinguish between `0` and `"0"` or `true` and `"true"`, they could do things like the following (though I think this would be quite rare):

```js
[ "let", "input", input,
  [ "match", ["concat", ["typeof", ["var", "input"]], "-", ["to_string", ["var", "input"]]],
    "string-0", e_a: R,
    "number-0", e_b: R,
    ...
  ]
]  
    
  
```